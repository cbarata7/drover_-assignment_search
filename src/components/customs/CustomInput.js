import React, { Component } from 'react';
import { FormControl } from 'react-bootstrap';

/**
* The custom input component
*/
class CustomInput extends Component {

  /**
  * Constructor that initializes the component
  * and sets its state
  */
  constructor () {
    super();

    this.state = {
      location: null,
      inputValue: ''
    }
  }

  /**
  * Function that handles the change of an input field
  * and calls the callback function from the parent component
  *
  * @param selectedOption the selectedOption of the inputed text
  */
  handleChange = (selectedOption) => {
    let newValue = {};

    newValue[this.props.metadataKey] = selectedOption.value;
    this.props.onBlurCustomInput(selectedOption);
  }

  /**
  * Function that handles the pressing of the enter key
  * and calls the input change funtion
  *
  * @param e the event object
  */
  _handleKeyPress = (e) => {
    if (e.key === 'Enter') {
      this.inputChange();
    }
  }

  /**
  * Function that handles the typping changes of the text value in the input field
  *
  * @param e the event object
  */
  handleText = (e) => {
    this.setState( {inputValue: e.target.value} );
  }

  /**
  * Function that handles the final changes of the text value in the input field
  */
  inputChange = () => {
    this.handleChange( {location: this.state.inputValue} );
  }

  /**
  * Function that renders the JSX to the custom input field
  */
  render () {
    return <div className='side-bar-option'>
      <div className='row side-bar-labels'>
        { this.props.name }
      </div>
      <div className='row'>
        <FormControl
          type='text'
          ref={ this.state.inputValue }
          onChange={ this.handleText }
          onBlur={ this.inputChange }
          defaultValue={ this.props.defaultOption }
          onKeyPress={ this._handleKeyPress }
          />
      </div>
    </div>
  }
}

export default CustomInput
