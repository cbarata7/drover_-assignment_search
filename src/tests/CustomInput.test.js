import React from 'react';
import { shallow } from 'enzyme';
import CustomInput from '../components/customs/CustomInput';

const clickFn = jest.fn();

describe('CustomInput', () => {
  it('should render correctly in debug mode', () => {
    const component = shallow(<CustomInput debug />);

    expect(component).toMatchSnapshot();
  });

  it('should render correctly with no props', () => {
    const component = shallow(<CustomInput/>);

    expect(component).toMatchSnapshot();
  });

  it('should render banner text correctly with given props', () => {
    const name = 'test';
    const metadataKey = 'test_key';
    const defaultOption = 'test label';
    const onBlurCustomInput = () => { return 'success!'};

    const component = shallow(
      <CustomInput
        name={name}
        metadataKey={metadataKey}
        defaultOption={defaultOption}
        onBlurCustomInput={onBlurCustomInput}/>
    );
    expect(component).toMatchSnapshot();
  });

  it('input blur should call the callback function with the value', () => {
    const name = 'test';
    const metadataKey = 'test_key';
    const defaultOption = 'test label';
    const onBlurCustomInput = () => { return 'success!'};

    const component = shallow(
      <CustomInput
        name={name}
        metadataKey={metadataKey}
        defaultOption={defaultOption}
        onBlurCustomInput={clickFn}/>
    );
    component
    .find('FormControl')
    .simulate('change', { target: { value: 'London' } })
    .simulate('blur');
    expect(clickFn).toHaveBeenCalledWith({'location': 'London'});
  });

  it('input change should change the state', () => {
    const name = 'test';
    const metadataKey = 'test_key';
    const defaultOption = 'test label';
    const onBlurCustomInput = () => { return 'success!'};

    const component = shallow(
      <CustomInput
        name={name}
        metadataKey={metadataKey}
        defaultOption={defaultOption}
        onBlurCustomInput={clickFn}/>
    );
    component
    .find('FormControl')
    .simulate('change', { target: { value: 'London' } });

    expect(component.state().inputValue).toEqual('London');
  });

  it('input blur should call the callback function with the value', () => {
    const name = 'test';
    const metadataKey = 'test_key';
    const defaultOption = 'test label';
    const onBlurCustomInput = () => { return 'success!'};

    const component = shallow(
      <CustomInput
        name={name}
        metadataKey={metadataKey}
        defaultOption={defaultOption}
        onBlurCustomInput={clickFn}/>
    );
    component
    .find('FormControl')
    .simulate('change', { target: { value: 'London' } })
    .simulate('blur');
    expect(clickFn).toHaveBeenCalledWith({'location': 'London'});
  });

});
