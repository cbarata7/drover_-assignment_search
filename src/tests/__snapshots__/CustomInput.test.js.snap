// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomInput should render banner text correctly with given props 1`] = `
<div
  className="side-bar-option"
>
  <div
    className="row side-bar-labels"
  >
    test
  </div>
  <div
    className="row"
  >
    <FormControl
      bsClass="form-control"
      componentClass="input"
      defaultValue="test label"
      onBlur={[Function]}
      onChange={[Function]}
      onKeyPress={[Function]}
      type="text"
    />
  </div>
</div>
`;

exports[`CustomInput should render correctly in debug mode 1`] = `
<div
  className="side-bar-option"
>
  <div
    className="row side-bar-labels"
  />
  <div
    className="row"
  >
    <FormControl
      bsClass="form-control"
      componentClass="input"
      onBlur={[Function]}
      onChange={[Function]}
      onKeyPress={[Function]}
      type="text"
    />
  </div>
</div>
`;

exports[`CustomInput should render correctly with no props 1`] = `
<div
  className="side-bar-option"
>
  <div
    className="row side-bar-labels"
  />
  <div
    className="row"
  >
    <FormControl
      bsClass="form-control"
      componentClass="input"
      onBlur={[Function]}
      onChange={[Function]}
      onKeyPress={[Function]}
      type="text"
    />
  </div>
</div>
`;
